### PR TITLE
Trilinos: support cuda, hwloc, Kokkos_ARCH_* and c++ std

### PR DESCRIPTION
**Pull request by Char-Aznable**
@keitat @aprokop This should address most of your concerns in https://github.com/spack/spack/pull/17900 Spack is new to me so suggestion is welcome

### CODE REVIEW

**Review by aprokop — COMMENTED** [review]
fwiw, it looks good to me. I don't think there is a way to avoid duplicating architecture lists from the Kokkos package, not without Kokkos being upstream of Trilinos.

It's hard to say which versions of Trilinos it will work with, though.

**Comment by aprokop** [comment]
This needs signoff by @balay and @keitat.

**Comment by sethrj** [comment]
@Char-Aznable did you happen to compare to another pull request that's adding CUDA support, https://github.com/spack/spack/pull/17900 ?

cc @amklinv-nnl @keitat

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/trilinos/package.py]:
```suggestion
        options.append(self.define_from_variant("std", "CMAKE_CXX_STANDARD"))
```

**Comment by keitat** [comment]
> @Char-Aznable did you happen to compare to another pull request that's adding CUDA support, #17900 ?
> 
> cc @amklinv-nnl @keitat
@Char-Aznable 
It is not working for me, trying to get more error info.  I need comments from @amklinv-nnl.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/trilinos/package.py]: I don't think `cuda` should default to on (on systems that support it), especially since CUDA support is available for only a small fraction of trilinos packages.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/trilinos/package.py:81]: Use single quotes for consistency here and with the other variants/conflicts.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/trilinos/package.py]: You should be able to do:
```python
from spack.pkg.builtin.kokkos import Kokkos
```

and use `Kokkos.spack_micro_arch_map` instead. Also note that your symbol here `kokkosArchCPU` is not PEP-8 compliant.

**Review by sethrj — CHANGES_REQUESTED** [review]
Needs some changes, and given the trouble that others had compiling cmake, it would be good if you could provide the full spec of the configuration for which you've successfully installed cuda.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (amklinv-nnl) [reply var/spack/repos/builtin/packages/trilinos/package.py]: Doesn't the cuda variant get inherited from CudaPackage anyway?

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (amklinv-nnl) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Shouldn't there be a conflict if the wrapper is not being used, but cuda is enabled and the compiler is not clang?

Comment (Char-Aznable) [reply var/spack/repos/builtin/packages/trilinos/package.py]: what's the correct spack syntax for `when='~%clang'`?

**Comment by Char-Aznable** [comment]
> Needs some changes, and given the trouble that others had compiling cmake, it would be good if you could provide the full spec of the configuration for which you've successfully installed cuda.

I don't know what this means but I have successfully installed trilinos+cuda on both centos 7 and ubuntu 18. Here is my trilinos spec:
```
trilinos@develop%gcc@9.3.0  ~adios2~alloptpkgs~amesos~amesos2~anasazi~aztec~belos+boost~cgns~chaco~complex+cuda~debug~dtk~epetra~epetraext~exodus+explicit_template_instantiation+float~fortran~glm+gtest+hdf5+hwloc~hypre~ifpack~ifpack2~intrepid~intrepid2~isorropia+kokkos~matio~mesquite~metis~minitensor~ml+mpi~muelu~mumps~netcdf~nox+openmp~phalanx~piro~pnetcdf~python+rol~rythmos~sacado~shards+shared~shylu~stk~stratimikos+suite-sparse~superlu~superlu-dist~teko~tempus+teuchos+tpetra+wrapper~x11~xsdkflags~zlib~zoltan~zoltan2 build_type=Release cuda_arch=61 gotype=long std=14 arch=linux-centos7-haswell
    ^boost@1.74.0%gcc@9.3.0  +atomic+chrono~clanglibcpp~container~context~coroutine+date_time~debug+exception~fiber+filesystem+graph~icu+iostreams+locale+log+math~mpi+multithreaded~numpy~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave cxxstd=98 visibility=hidden arch=linux-centos7-haswell
        ^bzip2@1.0.8%gcc@9.3.0  +shared arch=linux-centos7-haswell
            ^diffutils@3.7%gcc@9.3.0   arch=linux-centos7-haswell
                ^libiconv@1.16%gcc@9.3.0   arch=linux-centos7-haswell
        ^zlib@1.2.11%gcc@9.3.0  +optimize+pic+shared arch=linux-centos7-haswell
    ^cmake@3.18.2%gcc@9.3.0  ~doc+ncurses+openssl+ownlibs~qt arch=linux-centos7-haswell
        ^ncurses@6.2%gcc@9.3.0  ~symlinks+termlib arch=linux-centos7-haswell
            ^pkgconf@1.7.3%gcc@9.3.0   arch=linux-centos7-haswell
        ^openssl@1.1.1g%gcc@9.3.0  +systemcerts arch=linux-centos7-haswell
            ^perl@5.30.3%gcc@9.3.0  +cpanm+shared+threads arch=linux-centos7-haswell
                ^berkeley-db@18.1.40%gcc@9.3.0   arch=linux-centos7-haswell
                ^gdbm@1.18.1%gcc@9.3.0   arch=linux-centos7-haswell
                    ^readline@8.0%gcc@9.3.0   arch=linux-centos7-haswell
    ^cuda@11.0.2%gcc@9.3.0   arch=linux-centos7-haswell
        ^libxml2@2.9.10%gcc@9.3.0  ~python arch=linux-centos7-haswell
            ^xz@5.2.5%gcc@9.3.0  ~pic arch=linux-centos7-haswell
    ^hdf5@1.10.7%gcc@9.3.0  ~cxx~debug~fortran+hl~java+mpi+pic+shared~szip~threadsafe api=none arch=linux-centos7-haswell
        ^openmpi@4.0.5%gcc@9.3.0  ~atomics+cuda+cxx+cxx_exceptions~gpfs~java~legacylaunchers~lustre~memchecker~pmi~singularity~sqlite3+static+thread_multiple+vt+wrapper-rpath fabrics=none schedulers=sge arch=linux-centos7-haswell
            ^hwloc@2.2.0%gcc@9.3.0  ~cairo+cuda~gl~libudev+libxml2~netloc+nvml+pci+shared arch=linux-centos7-haswell
                ^libpciaccess@0.16%gcc@9.3.0   arch=linux-centos7-haswell
                    ^libtool@2.4.6%gcc@9.3.0   arch=linux-centos7-haswell
                        ^m4@1.4.18%gcc@9.3.0  +sigsegv patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 arch=linux-centos7-haswell
                            ^libsigsegv@2.12%gcc@9.3.0   arch=linux-centos7-haswell
                    ^util-macros@1.19.1%gcc@9.3.0   arch=linux-centos7-haswell
            ^numactl@2.0.12%gcc@9.3.0   arch=linux-centos7-haswell
                ^autoconf@2.69%gcc@9.3.0   arch=linux-centos7-haswell
                ^automake@1.16.2%gcc@9.3.0   arch=linux-centos7-haswell
    ^kokkos-nvcc-wrapper@3.2.00%gcc@9.3.0  +mpi arch=linux-centos7-haswell
    ^openblas@0.3.10%gcc@9.3.0  ~consistent_fpcsr~ilp64+pic+shared threads=openmp arch=linux-centos7-haswell
    ^suite-sparse@5.7.2%gcc@9.3.0  ~cuda~openmp+pic~tbb arch=linux-centos7-haswell
        ^metis@5.1.0%gcc@9.3.0  ~gdb~int64~real64+shared build_type=Release patches=4991da938c1d3a1d3dea78e49bbebecba00273f98df2a656e38b83d55b281da1,b1225da886605ea558db7ac08dd8054742ea5afe5ed61ad4d0fe7a495b1270d2 arch=linux-centos7-haswell
```

> > It is not working for me, trying to get more error info.  I need comments from @amklinv-nnl.
@keitat what's the  error message you got?


> @Char-Aznable did you happen to compare to another pull request that's adding CUDA support, #17900 ?
> 
Yes, it has most of what #17900 has except for the config of a few subpackage variants of Trilinos that I am not familiar with setting.

**Comment by Char-Aznable** [comment]
I have incorporated most of your suggestions. But I am not sure how to declare a conflicts when the compiler is not clang -- what's the correct spack syntax for when='~%clang'?

**Comment by amklinv-nnl** [comment]
> Yes, it has most of what #17900 has except for the config of a few subpackage variants of Trilinos that I am not familiar with setting.

My PR included the ability to run Trilinos tests, because I wanted to make sure that my changes weren't breaking Trilinos somehow. A lot of the changes in my PR related to failure to compile tests.

**Comment by Char-Aznable** [comment]
Are spack packages tested by CI/CD pipeline? I am not sure adding tests here makes sense to other people. I can add them if it does

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [reply var/spack/repos/builtin/packages/trilinos/package.py]: Yeah, the difference is that by overriding it, a package can specify it to be *on* by default (or in this case, on when building on non-mac systems).

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [reply var/spack/repos/builtin/packages/trilinos/package.py]: There isn't one. If `kokkos-nvcc-wrapper` is incompatible with clang then the restriction should be moved down into that.

**Comment by sethrj** [comment]
@Char-Aznable Spack CI doesn't test the validity of recipes (although I think there are efforts underway for "smoke tests" that will verify to first order xSDK/E4S installations). I notice that your spec has almost nothing enabled except for `+tpetra+suite-sparse`. Can you try a build with the default trilinos variants (make sure your packages.yaml isn't disabling them), plus cuda? The idea being that someone who tries `spack install trilinos+cuda` should either get a spack conflict message or a working build.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Are  you sure about this? I got build errors when building with `+amesos` and newer versions of superlu.

Comment (Char-Aznable) [reply var/spack/repos/builtin/packages/trilinos/package.py]: This one is in your official release: https://github.com/spack/spack/blob/4d586c598b758210608a9cf6275cd0f1ec10d3e3/var/spack/repos/builtin/packages/trilinos/package.py#L355 I think I somehow pick that up when rebasing my local repo

Comment (Char-Aznable) [reply var/spack/repos/builtin/packages/trilinos/package.py:366]: > @Char-Aznable Spack CI doesn't test the validity of recipes (although I think there are efforts underway for "smoke tests" that will verify to first order xSDK/E4S installations). I notice that your spec has almost nothing enabled except for `+tpetra+suite-sparse`. Can you try a build with the default trilinos variants (make sure your packages.yaml isn't disabling them), plus cuda? The idea being that someone who tries `spack install trilinos+cuda` should either get a spack conflict message or a working build.

This PR already takes my machines quite a few hours to build and I currently can't afford building everything of Trilinos. Most importantly, if there is no public pipeline to test the build , then I am not sure if my assertion of certain spec builds successfully for me actually mean anything to others. Someone can always jump in saying that this doesn't work for them. Not that I am not willing to do such test locally but I want to know what the end is for this.

**Comment by Char-Aznable** [comment]
>  @Char-Aznable Spack CI doesn't test the validity of recipes (although I think there are efforts underway for "smoke tests" that will verify to first order xSDK/E4S installations). I notice that your spec has almost nothing enabled except for +tpetra+suite-sparse. Can you try a build with the default trilinos variants (make sure your packages.yaml isn't disabling them), plus cuda? The idea being that someone who tries spack install trilinos+cuda should either get a spack conflict message or a working build.

This PR already takes my machines quite a few hours to build and I currently can't afford building everything of Trilinos. Most importantly, if there is no public pipeline to test the build , then I am not sure if my assertion of certain spec builds successfully for me actually mean anything to others. Someone can always jump in saying that this doesn't work for them. Not that I am not willing to do such test locally but I want to know what the end is for this.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (Char-Aznable) [reply var/spack/repos/builtin/packages/trilinos/package.py]: This is now removed

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (Char-Aznable) [reply var/spack/repos/builtin/packages/trilinos/package.py]: OK, I guess there is no sensible way to implement @amklinv-nnl 's suggestion then

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Sorry, I should have caught this earlier, but the standard Spack variant for C++ standard is `cxxstd` (`"variant('.*std'" */package.py`). Could you change it and line 733?

**Comment by sethrj** [comment]
> This PR already takes my machines quite a few hours to build and I currently can't afford building everything of Trilinos. Most importantly, if there is no public pipeline to test the build , then I am not sure if my assertion of certain spec builds successfully for me actually mean anything to others. Someone can always jump in saying that this doesn't work for them. Not that I am not willing to do such test locally but I want to know what the end is for this.

This then is definitely a good reason not to enable new features (CUDA) by default. There are ongoing discussions about how to make Trilinos easier and faster to build, so hopefully this use case will push in that direction.

I had one last comment about something that I noticed; thanks for addressing the others and adding CUDA support!

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (Char-Aznable) [reply var/spack/repos/builtin/packages/trilinos/package.py]: Fixed now

**Comment by Char-Aznable** [comment]
@sethrj I have rebased the PR on the latest of develop. I keep picking up the line `    depends_on('superlu-dist@5.4:6.2.0', when='@12.12.2:13.0+superlu-dist')` during rebase so I'll leave it there since I am not familiar with that part of Trilinos.

**Comment by keitat** [comment]
FYI: Trilinos 13 will support superlu_dist 6.3.

**Comment by keitat** [comment]
The future version of Trilinos will require C++14 because of the requirements for Kokkos.

**Comment by Char-Aznable** [comment]
What do you mean by "The future version"? Do you have a concrete version number to add?

Also, how do people handle PR here? It seems you have a different pipeline for handling PR than that for direct commit

**Comment by keitat** [comment]
> What do you mean by "The future version"? Do you have a concrete version number to add?
> 
> Also, how do people handle PR here? It seems you have a different pipeline for handling PR than that for direct commit

 C++14 will be the requirement for Trilinos 13.0 (as Kokkos uses C++14  capabilities.).  I am handing the PR from the perspective of xSDK project, but other ppl here have better clarification for that.

**Comment by sethrj** [comment]
@Char-Aznable unit tests are still failing:
```
Error: invalid syntax in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/trilinos/package.py, line 718
```
and according to the style/validate suite something in the code is incompatible with python 2.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Don't ever catch `Exception` without rethrowing it: what exception did you see that requires this? Shouldn't you just wrap in `if '+wrapper' in spec`?

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/trilinos/package.py]: This is the line that's not compatible with python 2 (or < 3.6ish I think). Just concatenate the two strings.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Same here.

**Review by sethrj — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (Char-Aznable) [reply var/spack/repos/builtin/packages/trilinos/package.py]: This is in your official release of kokkos/package.py. I simply copy and paste it. I don't really know what exception is supposed to catch here. Please make concrete suggestion so that I don't have to guess what the expected behavior is.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (Char-Aznable) [reply var/spack/repos/builtin/packages/trilinos/package.py]: fixed

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (Char-Aznable) [reply var/spack/repos/builtin/packages/trilinos/package.py]: fixed

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [reply var/spack/repos/builtin/packages/trilinos/package.py]: Sorry @Char-aznable, I didn't realize this was a copy-paste issue. Do you know what happens if you leave it off? Or what exceptions the code raises and under what circumstances?

@jjwilke it looks like you authored the original lines in the kokkos recipe in 598c233f78 ; can you explain what the `kokkos_cxx` and exception here are for?

Comment (jjwilke) [reply var/spack/repos/builtin/packages/trilinos/package.py]: Nope. Not sure why I would have done that instead of checking for the "+wrapper" variant. Probably I was just trying to figure out a way to check for a particular package as a dependency (rather than a variant) and couldn't find a better way.

I actually can't believe this got approved and merged. Usually the Spack team flags awfulness like this and tells me a better way to do it.

Comment (Char-Aznable) [reply var/spack/repos/builtin/packages/trilinos/package.py]: @jjwilke Do you think this `self.spec.kokkos_cxx` is not necessary at all for dependent package? Do you have a fix?

Comment (jjwilke) [reply var/spack/repos/builtin/packages/trilinos/package.py]: The `kokkos_cxx` is necessary for dependent packages if people rely on `nvcc_wrapper`.  I think just checking for the "+wrapper" variant is good enough and getting rid of the exception.

Comment (sethrj) [reply var/spack/repos/builtin/packages/trilinos/package.py]: Actually I'm not sure if this block of code is needed here  at all: downstream codes that need to  use the `nvcc_wrapper` can use `self.spec["kokkos-nvcc-wrapper"].kokkos_cxx`; the code here allows them to use `self.spec["trilinos"].kokkos_cxx` which I don't think is useful. It's certainly not needed for the `setup-build_environment` code below.

Comment (jjwilke) [reply var/spack/repos/builtin/packages/trilinos/package.py]: `kokkos-nvcc-wrapper` won't be part of the spec if it's a non-CUDA build.

Comment (sethrj) [reply var/spack/repos/builtin/packages/trilinos/package.py]: Sure, but the same is true here, right? `spec['trilinos']` isn't going to have a `kokkos_cxx` attribute if it's not building with CUDA.

Comment (jjwilke) [reply var/spack/repos/builtin/packages/trilinos/package.py]: Sorry, the intended behavior is that `spec['kokkos'].kokkos_cxx` is always valid regardless. Entirely possible that it is not implemented correctly.  This is all intended to work exactly the same way the MPI compilers do in Spack.

Comment (sethrj) [reply var/spack/repos/builtin/packages/trilinos/package.py]: Never mind, I'm being a dum-dum; I forgot the `except`. Changing this to an `if '+wrapper' in spec: ... else: ...` is the way to go here.

Comment (Char-Aznable) [reply var/spack/repos/builtin/packages/trilinos/package.py]: OK. This is fixed now

**Review by sethrj — APPROVED** [review]
Thanks for the updates, this looks good to me now!

**Comment by Char-Aznable** [comment]
Are we still waiting for signoff by @balay and @keitat?

**Comment by sethrj** [comment]
I think that would be good. I'm not an official Trilinos developer or maintainer but once they approve I'll hit the merge button.

**Comment by sethrj** [comment]
@Char-Aznable I tried a minimal build of trilinos on your PR and immediately hit a configure error:
```
$ spack install trilinos+cuda~boost~hypre~mumps~metis~suite-sparse cuda_arch=35
...
  >> 617    CMake Error at packages/kokkos/cmake/kokkos_test_cxx_std.cmake:109 (MESSAGE):
     618      Invalid compiler for CUDA.  The compiler must be nvcc_wrapper or Clang, but
     619      compiler ID was GNU
```

I now see that your `+wrapper` variant conflicts when `~cuda` but `~wrapper` doesn't conflict with `+cuda`. I think you should eliminate the wrapper variant since it seems to be required if and only if CUDA is in use.

**Comment by Char-Aznable** [comment]
@sethrj We have discussed that here https://github.com/spack/spack/pull/19119#discussion_r499813724 `~wrapper` only conflict with `+cuda` when `~%clang` but `~%clang` isn't really a thing

**Comment by sethrj** [comment]
> @sethrj We have discussed that here [#19119 (comment)](https://github.com/spack/spack/pull/19119#discussion_r499813724) `~wrapper` only conflict with `+cuda` when `~%clang` but `~%clang` isn't really a thing

Ah, I understand this now. @adamjstewart I think you've got some experience with weird conflicts and whatnot. Have any idea?

**Comment by sethrj** [comment]
By the way, I successfully (no build errors, but I haven't tested it) installed
```
trilinos+cuda~boost~hypre~mumps~metis~suite-sparse+wrapper cuda_arch=35
```
which concretized
```
trilinos@13.0.0%gcc@8.3.1~adios2~alloptpkgs+amesos+amesos2+anasazi+aztec+belos~boost~cgns~chaco~complex+cuda~debug~dtk+epetra+epetraext+exodus+explicit_template_instantiation~float+fortran+glm+gtest+hdf5~hwloc~hypre+ifpack+ifpack2~intrepid~intrepid2~isorropia+kokkos+matio~mesquite~metis~minitensor+ml+mpi+muelu~mumps+netcdf~nox~openmp~phalanx~piro~pnetcdf~python~rol~rythmos+sacado~shards+shared~shylu~stk~stratimikos~suite-sparse~superlu~superlu-dist~teko~tempus+teuchos+tpetra+wrapper~x11~xsdkflags~zlib+zoltan+zoltan2 build_type=RelWithDebInfo cuda_arch=35 cxxstd=11 gotype=long arch=linux-rhel8-haswell
```

**Comment by adamjstewart** [comment]
> > @sethrj We have discussed that here [#19119 (comment)](https://github.com/spack/spack/pull/19119#discussion_r499813724) `~wrapper` only conflict with `+cuda` when `~%clang` but `~%clang` isn't really a thing
> 
> Ah, I understand this now. @adamjstewart I think you've got some experience with weird conflicts and whatnot. Have any idea?

I saw something clever recently in the `gaussian-src` package:
```python
    # All compilers except for pgi are in conflict:                             
    for __compiler in spack.compilers.supported_compilers():                    
        if __compiler != 'pgi':                                                 
            conflicts('%{0}'.format(__compiler),                                
                      msg='Gaussian can only be built with the PGI compiler')
```
You could prob manipulate this to emulate `~%clang`.

**Comment by sethrj** [comment]
Thanks @adamjstewart ! Failing that we could add a spack-based error in the `cmake_args` or somewhere so it doesn't show up as a configure error.

**Comment by Char-Aznable** [comment]
fixed

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**
